docs/design/Extension_Management.md

```diff
@@ -0,0 +1,147 @@
+# Extension Management
+
+Upgrade Assistant's extensibility provides the ability for users to extend it beyond its original design. As we've worked with developers making extensions, it's clear that there are some issues with discoverability and maintainability of custom extensions. Some of the major pain points:
+
+- **Version management:** Need to be able to specify which extension version a project should use
+- **Distribution** Need to be able easily get a specific version (or update existing ones) of an extension
+- **Consistency between runs** Should be able to register an extension so subsequent runs automatically use it
+
+Throughout this document, the term `workspace` will be used to refer to a solution or project that is being updated along with the state file Upgrade Assistant generates.
+
+## User Experience
+
+Extensions will now be something that are managed at a workspace level. This will allow users to have a consistent experience as well as be able to manage their extensions.
+
+Below are the commands to manage extensions:
+
+```
+upgrade-assistant extension list
+```
+
+This command will list current extensions applied to a project.
+
+```
+upgrade-assistant extension add --name [name] [--source [source]]
+```
+
+This command will allow users to add extensions to a project.
+
+```
+upgrade-assistant extension remove --name [name]
+```
+
+This command will remove an installed extension from a current workspace.
+
+```
+upgrade-assistant extension restore
+```
+
+This command will restore extensions when they are not available
+
+> This may be able to be done automatically, but initially, we'll have it be a separate command
+
+```
+upgrade-assistant extension feed list [--source [source]]
+```
+
+This command will allow someone to list extensions available on a feed
+
+### Extension Author Commands
+
+```
+upgrade-assistant extension feed publish --path [path] --source [source]
+```
+
+This command will allow extension authors to publish an extension.
+
+```
+upgrade-assistant extension feed remove --name [name] --source [source]
+```
+
+This command will allow extension authors to remove a published extension.
+
+```
+upgrade-assistant extension feed clean --source [source]
+```
+
+This command will allow extension authors to automatically clean up a feed (i.e. if a file has been removed from the storage, this will notify the author as well as remove it from the feed metadata).
+
+## Local Cache
```

Comment:
Do we think it's worth specing this out before we've landed on distribution? I think we get a bunch of this for free with some of the options (as you mention). Or are you thinking of this is an abstraction over the distribution?

Reply:
This section deals more with how UA will track within a workspace what extensions are being used. I renamed the section and removed the part about where we could cache the extension as that doesn't need to be discussed until we land on a distribution system.